### ISSUE TITLE
Add Segment Trees

### ISSUE THREAD

**Issue by czgdp1807** [codezonediitj/pydatastructs#6]
#### Description of the problem

Add segment trees to `pydatastructs.trees.space_partitioning_trees`.

#### References/Other comments
Use, https://en.wikipedia.org/wiki/Segment_tree

**Comment by czgdp1807** [comment]
I am working on it.

**Comment by Smit-create** [comment]
Can we also add update operation? As I think this is the most necessary feature for which segment tree is used as it can update the interval list in O(logN) time. That is segment tree are generally used to solve online problems (real time update and queries). Moreover, without this update operation, quering on just given set of segments there exits O(1) technique(using prefix sums which build in O(MAXN) and each query in O(1))

**Comment by czgdp1807** [comment]
> It is, in principle, a static structure; that is, it's a structure that cannot be modified once it's built. A similar data structure is the interval tree. 

From wikipedia. So, for dynamic segment trees we should subclass `SegmentTree` in `DynamicSegmentTree` and add an update method to it. Another question is how will the algorithm for updating segment tree will work, especially the ones defined [here](https://en.wikipedia.org/wiki/Segment_tree).

**Comment by aman503** [comment]
Can I work on this issue?

**Comment by Rajveer100** [comment]
I can work on the issue..
I have so many templates..you can check in my GitHub..C++ -> Python I will do the implementation and conversion..

**Comment by czgdp1807** [comment]
Sure. Please read the above comments before starting.

**Comment by Rajveer100** [comment]
Thank you..will keep you updated with the progress...and will ask if you if any doubts are there regarding the same..

**Comment by Rajveer100** [comment]
For segment tree : 

<Build Tree> followed by :

1: Range Query 
2: Range Update(Lazy propagation)
3: Point Update

And common range functions like Sum, Min, Max, Gcd, etc..

Is there anything else I must add?
Or any other suggestions?

**Comment by czgdp1807** [comment]
@Rajveer100 Please feel free to open a PR. I will review there. Thanks.

**Comment by Rajveer100** [comment]
Yes I have already done the PR(483)..Thank you so much for your reply..
